### PR TITLE
[WIP]Dedicated machines

### PR DESCRIPTION
**Pull request by kevin-wangzefeng**
ref #17190

### CODE REVIEW

**Comment by kevin-wangzefeng** [comment]
@davidopp PTAL

**Comment by smarterclayton** [comment]
@derekwaynecarr FYI overlap with the up streaming node selector admission controller, want to make sure this covers use case

**Comment by derekwaynecarr** [comment]
@smarterclayton - will review, thanks for heads up

**Comment by derekwaynecarr** [comment]
I will do a deeper analysis tomorrow, but this seems problematic if there are more than one DedicatedMachine object per Namespace, is the idea that that would be restricted by quota so you get deterministic results?

**Comment by derekwaynecarr** [comment]
Major difference I see is that in OpenShift, a nodeSelector can be specified for the namespace that gets merged with the pods specified node selector.  So for example, the namespace node selector can say "infra=false,env=test", a pod could be created that has a node selector of "gpu=x".   The OpenShift use case will then verify that the pod node selector and the namespace node selector do not conflict.  If they do, it errors.  If they don't, the pods node selector is merged with the namespace selector to say "infra=false,env=test,gpu=x", and that is the scheduling requirement that is fulfilled.

I think the PR as shown here is too restrictive and should allow for merged node selectors.

More detail review will have to wait to tomorrow.

**Comment by davidopp** [comment]
I'll take a look at this within the next few days. I'm currently discussing the dedicated machine issue with @bgrant0607 .

**Comment by kevin-wangzefeng** [comment]
@derekwaynecarr thanks for review, I agree with you about the PR as shown here is too restrictive. 

For the case there are more than one DedicatedMachine object per Namespace, I don't have a good idea to solve it yet.

For merged node selectors, it is easy to do, I can update the DedicatedMachineSpec definition to:

```go
type DedicatedMachineSpec struct {
	NodeSelector map[string]string `json:"nodeSelector"`
}
```

**Comment by kevin-wangzefeng** [comment]
And @davidopp mentioned this can be done with annotations instead of labels, I think the overall structure maybe similar, just need to make scheduler to support annotations.

**Comment by ncdc** [comment]
cc @aveshagarwal

File: contrib/completions/bash/kubectl [hunk 1]
```diff
@@ -288,6 +288,7 @@ _kubectl_get()
     must_have_one_noun=()
     must_have_one_noun+=("componentstatus")
     must_have_one_noun+=("daemonset")
+    must_have_one_noun+=("dedicatedmachine")
```
Comment (bgrant0607) [inline contrib/completions/bash/kubectl]: We don't use the term "machine" anywhere in Kubernetes. Let's not introduce another term in addition to minion and host. Node is the appropriate term.

File: pkg/apis/extensions/types.go [hunk 2]
```diff
@@ -683,3 +683,30 @@ const (
...
+	// Standard list metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	unversioned.ListMeta `json:"metadata,omitempty"`
+
+	// Items is a list of DedicatedMachines.
+	Items []DedicatedMachine `json:"items"`
+}
+
+type DedicatedMachineSpec struct {
+	// TargetNamespace string `json:"namespace,omitempty"`
```
Comment (bgrant0607) [inline pkg/apis/extensions/types.go]: Comment doesn't match field.

Before making an API proposal, please read the documentation about how to do so. Also, it's a good idea to start with a proposal doc, in general, in order to minimize wasted effort.

https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/coding-conventions.md
https://github.com/kubernetes/kubernetes/blob/master/docs/api.md
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api_changes.md

**Comment by bgrant0607** [comment]
Please create a proposal first.

**Comment by derekwaynecarr** [comment]
:100: to proposal, it was hard to discern behavior when going right to the code.

As you write the proposal, please account for the following:

* Allow for different behaviors; i.e. merge, overwrite, etc. around the namespace node selector and pod node selector interaction.  Merge with validation on admission was critical for our downstream use case to be satisfied appropriately.
* What happens if there are many of these things, or if the name should be fixed and we only allow 1 per Namespace.
* What is the user expectation if I change the selector behavior, or add it once existing pods are in the system.  For example, while I do not expect users must have their existing pods rescheduled per se, I could imagine a CLI command that makes it easy to enumerate what pods in a namespace are not matching that namespace associated selector.

Look forward to proposal!

**Comment by kevin-wangzefeng** [comment]
@bgrant0607 @derekwaynecarr thanks for the feedback, I will create a proposal first.

**Comment by davidopp** [comment]
We have an alternative proposal, see #18263

**Pull request closed without merging**